t/future.t

```diff
@@ -0,0 +1,197 @@
+        my ($self) = @_;
+        return Future->fail('failure message', 'rpc');
+    };
+
+    provide deferred_fail => sub {
+        my ($self) = @_;
+        my $f = Future::Mojo->new;
+        Mojo::IOLoop->timer(0.5 => sub { $f->fail('deferred failure') });
+        return $f;
+    };
+
+    provide deferred_success => sub {
+        my ($self) = @_;
+        my $f = Future::Mojo->new;
+        Mojo::IOLoop->timer(0.5 => sub { $f->done('deferred ok') });
+        return $f;
+    };
+
+}
+
+#-------------------------------------------------------------------
+
+{
+# Mojolicious app for testing
+    package MojoxJsonRpc;
+
+    use Mojo::Base 'Mojolicious';
+
+    use MojoX::JSON::RPC::Service;
+# This method will run once at server start
+    sub startup {
+        my $self = shift;
+
+        $self->secrets(['Testing!']);
+
+        $self->plugin('json_rpc_dispatcher', services => {'/jsonrpc' => MyService->new});
+    }
+
+}
+
+#-------------------------------------------------------------------
+
+# Back to tests
+use TestUts;
+
+use Test::Mojo;
+
+my $t = Test::Mojo->new('MojoxJsonRpc');
+my $client = MojoX::JSON::RPC::Client->new(ua => $t->ua);
+
+note 'simple echo';
+TestUts::test_call(
+    $client,
+    '/jsonrpc',
+    {
+        id     => 2,
+        method => 'echo',
+        params => ['HEEEEEEEEEEEEEEEEELLLLLLLLLLLLLLLOOOOOOOOOOO!']
+    },
+    {
+        result => 'HEEEEEEEEEEEEEEEEELLLLLLLLLLLLLLLOOOOOOOOOOO!',
+        id     => 2
+    },
+    'echo 1'
+);
+
+# test Future done
+
+note 'immediate success';
+TestUts::test_call(
+    $client,
+    '/jsonrpc',
+    {
+        id     => 1,
+        method => 'immediate_success',
+    },
+    {
+        result => 'future done',
+        id     => 1
+    },
+    'future'
+);
+
+note 'deferred success, should take half a second';
+TestUts::test_call(
+    $client,
+    '/jsonrpc',
+    {
+        id     => 1,
+        method => 'deferred_success',
+    },
+    {
+        result => 'deferred ok',
+        id     => 1
+    },
+    'deferred success future'
+);
+
+TestUts::test_call(
+    $client,
+    '/jsonrpc',
+    {
+        id     => 1,
+        method => 'immediate_fail',
+    },
+    {
+        error => {
+            message => 'failure message',
+            code    => '',
+            data    => ''
+        },
+        id => 1
+    },
+    'immediate_fail'
+);
+
+TestUts::test_call(
+    $client,
+    '/jsonrpc',
+    {
+        id     => 1,
+        method => 'deferred_fail',
+    },
+    {
+        error => {
+            message => 'deferred failure',
+            code    => '',
+            data    => ''
+        },
+        id => 1
+    },
+    'deferred_fail'
+);
+
+TestUts::test_call(
+    $client,
+    '/jsonrpc',
+    {
+        id     => 2,
+        method => 'echo',
+        params => ['HEEEEEEEEEEEEEEEEELLLLLLLLLLLLLLLOOOOOOOOOOO!']
+    },
+    {
+        result => 'HEEEEEEEEEEEEEEEEELLLLLLLLLLLLLLLOOOOOOOOOOO!',
+        id     => 2
+    },
+    'echo 2'
+);
```

Comment:
I guess the idea behind repeating this test is to show that both synchronous responses and deferred ones work together. Could you please add a note why you are doing this? Otherwise future refactoring might remove it.